src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const eachAction = [];
+  let initialState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...initialState };
+
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+
+      case 'clear':
+        newState = {};
+        break;
+    }
+    eachAction.push(newState);
+    initialState = newState;
```

Comment:
[DONT REPEAT YOURSELF] - You are pushing the `newState` to the `eachAction` array and updating the `initialState` inside the switch block. To adhere to the DRY principle, consider performing these actions once after the switch block.